### PR TITLE
[Feature] Virtual Width Network Eagle3

### PR DESCRIPTION
**Pull request by YzTongNiar**
### What this PR does / why we need it?
We add a customed speculative model "Virtual Width Network Eagle3" with architecture similar to Seed's work 'Virtual Width Network' (https://huggingface.co/papers/2511.11238)  

We impelmented and trained the 'vwn eagle3' based on 'speculators' library  (https://github.com/vllm-project/speculators); the training datasets includes 'ultrachat' (https://huggingface.co/datasets/openbmb/UltraChat) and 'sharegpt' (https://huggingface.co/collections/bunnycore/sharegpt-datasets).

We used the same setting to train a vwn-eagle3 and an eagle3 model to compare their performace on longbenchv2, gsm8k, sharegpt, and humaneval; as well as some man-made fixed-length datasets. The performance is concluded as following:

Datasets testing result:

| Dataset | Context-Length | E2E Throughput Gain|
| ------ | ------ | ------ | 
| gsm8k   | 1.5k | 7.9%|
| humaneval | 2.5k | 11.94%|
| sharegpt  | 5k | 2.84%|
| longbench_v2| 128k | 18.23%|

Fixed length testing results:

|Input Lenght | Output Length | E2E Throughput Gain|
| ------ | ------ | ------ | 
| 2k   | 2k | 9.83%|
| 3.5k | 1.5k | 2.38%|
| 32k  | 1k | 18.66%|

### Does this PR introduce _any_ user-facing change?

Nope. VWN Eagle3 uses the same interface with eagle3 and can be distinguished by draft model's config file, thus, the user can simply use the same setting to start the server:
 `--speculative-config '{"method": "eagle3","model": "/vwn-eagle3-model-path/", "num_speculative_tokens": 3}' `

To reduce redundant codes, we reuse the implementation of llama_eagle3.py to the greatest extent：

  | Component | Provided by Upstream Eagle3 | Overridden/Added by VWN | Reuse Rate |
  |-----------|-----------------------------|------------------------|------------|
  | Top-level CausalLM | ✅  forward, embed_input_ids, load_weights | Only replaces `self.model` | ~80% |
  | Model container | ✅ embed, fc, load_weights  | Swaps layer type + simplifies forward | ~50% |
  | Decoder Layer | ✅  self_attn, mlp,  quant | Adds VWN projections + rewrites forward | ~40% |
  | PreVwnLayerV1 | ❌ No counterpart | Entirely new implementation | 0% (pure addition) |

### How was this patch tested?
- vLLM-ascend main @ 7cb4484d3b2e00733764eb6f17b4205723b925c7

- vLLM version: v0.22.1
- vLLM main: https://github.com/vllm-project/vllm/commit/967c5c3bc38891f4465d3f4e99917ed837bb3833

### CODE REVIEW

File: vllm_ascend/models/llama_eagle3_vwn.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        return self.norm(hidden_states, residual), hidden_states
+
+
+class Eagle3VwnLlamaForCausalLM(Eagle3LlamaForCausalLM):
+
+    def __init__(self, *, vllm_config, prefix=""):
+        super().__init__(vllm_config=vllm_config, prefix=prefix)
+        n = vllm_config.model_config.get_num_layers(vllm_config.parallel_config)
+        self.model = VwnLlamaModel(vllm_config=vllm_config, prefix="model",
+                                   start_layer_id=n)
```
Comment (dsxsteven) [inline vllm_ascend/models/llama_eagle3_vwn.py]: Please add ut for this new file

File: vllm_ascend/models/llama_eagle3_vwn.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        input_size=inp, output_size=out, bias=False,
+        params_dtype=vc.model_config.dtype,
+        quant_config=qc, prefix=pfx, return_bias=False)
+
+
+class PreVwnLayerV1(nn.Module):
+    def __init__(self, vllm_config, prefix="", config=None, quant_config=None):
+        super().__init__()
+        cfg = config or vllm_config.model_config.hf_config
+        hs, m, r = cfg.hidden_size, getattr(cfg, "vwn_m", 1), getattr(cfg, "vwn_r", 1)
```
Comment (dsxsteven) [inline vllm_ascend/models/llama_eagle3_vwn.py:40]: The default fallback 1 is safe, but if a config explicitly sets vwn_m = 0, every // m operation (lines 37, 52, 68, 71, 74, 76, 78) will raise ZeroDivisionError. There is no validation.You can Add an explicit guard if m/r <= 0;

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 3]
```diff
@@ -612,7 +613,10 @@ def _propose(
 
         if self.method in ("eagle3", "dflash"):
             assert isinstance(
-                self.get_model(), (Eagle3LlamaForCausalLM, DFlashQwen3ForCausalLM, Eagle3DeepseekV2ForCausalLM)
+                self.get_model(), (Eagle3LlamaForCausalLM,
+                                                 DFlashQwen3ForCausalLM,
+                                                 Eagle3VwnLlamaForCausalLM,
+                                                 Eagle3DeepseekV2ForCausalLM)
```
Comment (dsxsteven) [inline vllm_ascend/spec_decode/llm_base_proposer.py]: The indentation is inconsistent (extra indentation for DFlashQwen3ForCausalLM and Eagle3VwnLlamaForCausalLM). This should be aligned.

File: vllm_ascend/models/llama_eagle3_vwn.py (hunk 2 above)
Comment (YzTongNiar) [reply vllm_ascend/models/llama_eagle3_vwn.py:40]: thanks for your comments; since the parameter 'vwn_m' is loaded from the config file of the weight, I really do not expect any wrong parameters, except that someone modifies the config file.

File: vllm_ascend/models/llama_eagle3_vwn.py (hunk 1 above)
Comment (YzTongNiar) [reply vllm_ascend/models/llama_eagle3_vwn.py]: UT has been added

File: vllm_ascend/models/llama_eagle3_vwn.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                quant_config=qc,
+                prefix=maybe_prefix(prefix, "self_attn.qkv_proj"),
+            )
+
+        mp = maybe_prefix
+        self.pre_vwn_layer = PreVwnLayerV1(vllm_config, mp(prefix, "layers.pre_vwn_layer"), cfg, qc)
+        self.downward_and_forgot = _linear(wd // m, (hs + wd) // m, vllm_config, qc, mp(prefix, "downward_and_forgot"))
+        self.pre_attention_layernorm = RMSNorm(hs, eps=cfg.rms_norm_eps)
+        self.upward_after_attn = _linear(hs // m, wd // m, vllm_config, qc, mp(prefix, "upward_after_attn"))
+        self.downward_and_forgot_after_attn = _linear(
```
Comment (realliujiaxu) [inline vllm_ascend/models/llama_eagle3_vwn.py:78]: please check if `downward_and_forgot` should `downward_and_forgot_after_attn`

File: vllm_ascend/spec_decode/llm_base_proposer.py (hunk 3 above)
Comment (YzTongNiar) [reply vllm_ascend/spec_decode/llm_base_proposer.py]: The lint problem has been fixed.

File: vllm_ascend/models/llama_eagle3_vwn.py (hunk 4 above)
Comment (YzTongNiar) [reply vllm_ascend/models/llama_eagle3_vwn.py:78]: The 'Eagle3LlamaForCausalLM.load_weights' method uses  'AutoWeightsLoader' for weights loading, it matches the key of the weights (*.downward_and_forgot_after_attn.weight) with the model's argument (self.downward_and_forgot_after_attn). The prefix argument only affects quant configs and logging, **thus it acutally has no functional error by using two 'downward_and_forgot'. But for better understanding,I will use 'downward_and_forgot_after_attn' instead.**

File: tests/e2e/pull_request/two_card/spec_decode/test_spec_decode.py [hunk 5]
```diff
@@ -327,3 +335,95 @@ def test_p_eagle_acceptance(
...
+            num_drafts += metric.value
+        elif metric.name == "vllm:spec_decode_num_accepted_tokens_per_pos":
+            assert isinstance(metric, Vector)
+            for pos in range(len(metric.values)):
+                num_accepted_tokens_per_pos[pos] += metric.values[pos]
+
+    acceptance_per_pos = [n / num_drafts for n in num_accepted_tokens_per_pos]
+    golden = BASELINES_SP["vwn_eagle3"]
+
+    match = all(abs(a - b) < 0.1 for a, b in zip(acceptance_per_pos, golden))
```
Comment (lilinsiman) [inline tests/e2e/pull_request/two_card/spec_decode/test_spec_decode.py]: The allowed fluctuation of 10% is too large. It is recommended that the maximum fluctuation be 6% or less. As a result, the case cannot be used to monitor the acceptance rate.

Comment (YzTongNiar) [reply tests/e2e/pull_request/two_card/spec_decode/test_spec_decode.py]: has been modified to 6%

**Comment by YzTongNiar** [comment]
[Bot]: rerun completed.

**Rerun:**
- E2E

**Review by lilinsiman — APPROVED** [review]

**Review by zzzzwwjj — APPROVED** [review]
